source/opt/constants.cpp

```diff
@@ -0,0 +1,215 @@
+// Copyright (c) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "constants.h"
+#include "ir_context.h"
+
+#include <unordered_map>
+#include <vector>
+
+namespace spvtools {
+namespace opt {
+namespace analysis {
+
+analysis::Type* ConstantManager::GetType(const ir::Instruction* inst) const {
+  return context()->get_type_mgr()->GetType(inst->type_id());
+}
+
+uint32_t ConstantManager::FindRecordedConstant(
+    const analysis::Constant* c) const {
+  auto iter = const_val_to_id_.find(c);
+  if (iter == const_val_to_id_.end()) {
+    return 0;
+  } else {
+    return iter->second;
+  }
+}
+
+std::vector<const analysis::Constant*> ConstantManager::GetConstantsFromIds(
+    const std::vector<uint32_t>& ids) const {
+  std::vector<const analysis::Constant*> constants;
+  for (uint32_t id : ids) {
+    if (analysis::Constant* c = FindRecordedConstant(id)) {
+      constants.push_back(c);
+    } else {
+      return {};
+    }
+  }
+  return constants;
+}
+
+ir::Instruction* ConstantManager::BuildInstructionAndAddToModule(
+    std::unique_ptr<analysis::Constant> c, ir::Module::inst_iterator* pos) {
+  analysis::Constant* new_const = c.get();
+  uint32_t new_id = context()->TakeNextId();
+  const_val_to_id_[new_const] = new_id;
+  id_to_const_val_[new_id] = std::move(c);
+  auto new_inst = CreateInstruction(new_id, new_const);
+  if (!new_inst) return nullptr;
+  auto* new_inst_ptr = new_inst.get();
+  *pos = pos->InsertBefore(std::move(new_inst));
+  ++(*pos);
+  context()->get_def_use_mgr()->AnalyzeInstDefUse(new_inst_ptr);
```

Comment:
We should add something to the context to do this.  Then it can update the def use manager if it is built.  This will rebuild it even if it is not needed.  See `IRContext::AnalyzeUses` for how this could be done.